#!/bin/bash
# ==========================================
#   ATAC-seq Data Processing Pipeline
#   Author: Ruoxuan Wang
#   Date: 08/26/2025
#   Description: A general pipeline for ATAC-seq
# ==========================================

# -------- User-defined variables ----------

R1=sample_R1.fastq.gz
R2=sample_R2.fastq.gz
SAMPLE=sample1
GENOME_FA=/path/to/genome.fa
GENOME_INDEX=/path/to/genome_index  # Generated by bowtie2-build 
GENOME_SIZE=1.35e8                  # Arabidopsis: 1.35e8
THREADS=16
# ------------------------------------------


#############################
# Step 1. QC & Adapter Trim #
#############################

fastp \
  -i $R1 \
  -I $R2 \
  -o ${SAMPLE}_clean_R1.fq.gz \
  -O ${SAMPLE}_clean_R2.fq.gz \
  --detect_adapter_for_pe \
  -h ${SAMPLE}_fastp_report.html \
  -j ${SAMPLE}_fastp_report.json


##########################
# Step 2. Genome Mapping #
##########################

# (Optional, only once) Build genome index
# bowtie2-build $GENOME_FA $GENOME_INDEX

bowtie2 \
  -p $THREADS \
  -X 2000 \
  --very-sensitive \
  -x $GENOME_INDEX \
  -1 ${SAMPLE}_clean_R1.fq.gz \
  -2 ${SAMPLE}_clean_R2.fq.gz \
  | samtools view -bS - > ${SAMPLE}.bam


##################################
# Step 3. Sort & Add Read Groups #
##################################

samtools sort -o ${SAMPLE}.sorted.bam ${SAMPLE}.bam
samtools index ${SAMPLE}.sorted.bam

samtools addreplacerg \
  -r "@RG\tID:${SAMPLE}\tSM:${SAMPLE}\tPL:ILLUMINA" \
  -o ${SAMPLE}.rg.bam \
  ${SAMPLE}.sorted.bam

samtools index ${SAMPLE}.rg.bam


###############################
# Step 4. Remove Duplicates   #
###############################

picard MarkDuplicates \
  I=${SAMPLE}.rg.bam \
  O=${SAMPLE}.dedup.bam \
  M=${SAMPLE}_dupmetrics.txt \
  REMOVE_DUPLICATES=true

samtools index ${SAMPLE}.dedup.bam


##########################
# Step 5. Tn5 Shift Fix  #
##########################

alignmentSieve \
  -b ${SAMPLE}.dedup.bam \
  --ATACshift \
  -o ${SAMPLE}.shifted.bam

samtools sort -o ${SAMPLE}.shifted.sorted.bam ${SAMPLE}.shifted.bam
samtools index ${SAMPLE}.shifted.sorted.bam


######################
# Step 6. Peak Calling
######################

macs2 callpeak \
  -t ${SAMPLE}.shifted.sorted.bam \
  --nomodel \
  --shift -100 \
  --extsize 200 \
  -q 0.01 \
  -g $GENOME_SIZE \
  -f BAMPE \
  -n ${SAMPLE} \
  --outdir macs2_output/


##################################
# Step 7. Generate BigWig tracks #
##################################

bamCoverage \
  -b ${SAMPLE}.shifted.sorted.bam \
  -o ${SAMPLE}.bw \
  --normalizeUsing CPM \
  --binSize 10

echo "ATAC-seq pipeline finished for $SAMPLE"
